src/lib/optimizer/strategy/nested_expression_rule.cpp

```diff
@@ -0,0 +1,163 @@
+    if (!variant_is_null(value) &&
+        _replace_expression_in_parents(node, node->output_column_references()[column_id], value)) {
+      // If we successfully replaced the occurences of the Expression in the parent tree
+      // we can remove the column here.
+      _remove_column_from_projection(projection_node, column_id);
+    }
+  }
+
+  return _apply_to_children(node);
+}
+
+bool NestedExpressionRule::_replace_expression_in_parents(const std::shared_ptr<AbstractLQPNode>& node,
+                                                          const LQPColumnReference& expression_column,
+                                                          const AllTypeVariant& value) {
+  auto parent_tree_changed = false;
+  for (auto parent : node->parents()) {
+    if (parent->type() != LQPNodeType::Predicate) {
+      parent_tree_changed |= _replace_expression_in_parents(parent, expression_column, value);
+      continue;
+    }
+
+    auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(parent);
+
+    // We look for a PredicateNode which has an LQPColumnReference as value,
+    // referring to the column which contains the Expression we resolved before.
+    if (is_lqp_column_reference(predicate_node->value()) &&
+        boost::get<LQPColumnReference>(predicate_node->value()) == expression_column) {
+      // We replace the LQPColumnReference with the actual result of the Expression it was referring to.
+      auto new_predicate_node = std::make_shared<PredicateNode>(predicate_node->column_reference(),
+                                                                predicate_node->predicate_condition(), value);
+      predicate_node->replace_with(new_predicate_node);
+      parent_tree_changed = true;
+    }
+
+    parent_tree_changed |= _replace_expression_in_parents(parent, expression_column, value);
+  }
+  return parent_tree_changed;
+}
+
+void NestedExpressionRule::_remove_column_from_projection(const std::shared_ptr<ProjectionNode>& node,
+                                                          ColumnID column_id) {
+  auto column_expressions = node->column_expressions();
+  column_expressions.erase(column_expressions.begin() + column_id);
+
+  auto projection_node = std::make_shared<ProjectionNode>(column_expressions);
+  node->replace_with(projection_node);
+}
+
+std::optional<DataType> NestedExpressionRule::_get_type_of_expression(
+    const std::shared_ptr<LQPExpression>& expression) const {
+  if (expression->type() == ExpressionType::Literal) {
+    return data_type_from_all_type_variant(expression->value());
+  }
+
+  if (!expression->is_arithmetic_operator()) {
+    return std::nullopt;
+  }
+
+  const auto type_left = _get_type_of_expression(expression->left_child());
+  const auto type_right = _get_type_of_expression(expression->right_child());
+
+  if (type_left == DataType::Null) return type_right;
+  if (type_right == DataType::Null) return type_left;
+  if (type_left != type_right) return std::nullopt;
```

Comment:
`_get_type_of_expression` looks like it does two things at once. Getting the type and checking if the rule can be applied. Also, the return type is weird - in some cases, you use it, in others, you abuse `optional` to return true/false.

`_evaluate_expression` does its own checks (e.g., expression->is_arithmetic_operator) and does another recursion.

How about a single method `optional<AllTypeVariant> _calculate_expression` that returns `nullopt` if the rule cannot be applied?

Reply:
I have tried using only a single method for this for a bit right now, but I don't think I can avoid doing two recursions. I need to know the data type before I get into the `_calculate_expression` recursion, because otherwise I have no way of getting it, unless I want to go through the tree again every recursion of `_calculate_expression`.

However I have cut a lot of code from the method as you suggested further down, which makes it more understandable, thank you!

And yes, for both methods now, `_calculate_expression` and `_get_type_of_expression`, I am using `nullopt` to indicate that the rule cannot be applied.